exercises/joseph_ramsay/ex01_basics/src/main.cpp

```diff
@@ -0,0 +1,182 @@
+ */
+bool notLowerAlNum(const char c) {
+  return c < '0' || c > 'z' || (c > '9' && c < 'a');
+}
+
+/** Remove punctuation from a string
+ *
+ * @param s The string to remove from
+ * @return The modified string
+ */
+string noPunct(string s) {
+  string::iterator iter(remove_if(s.begin(), s.end(), notLowerAlNum));
+  s.erase(iter, s.end());
+  return s;
+}
+
+/** Get the maximum length of word in a word counter
+ * @param counter The word counter
+ * @return The length of the longest word
+ */
+int maxLen(const map<string, int> &counter) {
+  int m = 0;
+  for (const auto &element : counter){
+    m = max(m, int(element.first.length()));
+  }
+  return m;
+}
+
+/** Compare two pairs, according to their second element
+ * @param p1 The first pair
+ * @param p2 The second pair
+ * @return True if p1's second element is greater than p2's, false otherwise
+ */
+bool comparePair(const pair<string, int> &p1, const pair<string, int> &p2) {
+  return p1.second > p2.second;
+}
+
+/** Copy a map into a vector, sorted in descening order by the map's values
+ * @param counter The map to copy
+ * @return A sorted vector of pairs
+ */
+vector<pair<string, int>> sortedMap(map<string, int> &counter) {
+  vector<pair<string, int>> sorted;
+  copy(counter.cbegin(), counter.cend(), back_inserter(sorted));
+  //for_each(counter.begin(), counter.end(),
+  //         [&sorted](pair<string, int> p) { sorted.push_back(p); });
+  sort(sorted.begin(), sorted.end(), comparePair);
+  return sorted;
+}
+
+void displayError(string msg){
+  cout << msg << "\nUsage: ./WordCounter infile [outfile]\n";
+}
+
+
+int main(int argv, char **argc) {
+  if (argv <= 1) { // User didn't supply enough arguments
+    displayError("Please enter a filename");
+    return 1;
+  }
+
+  ifstream infile(argc[1]);
+
+  if (!infile.good()) { // User entered a bad file
+    displayError("File does not exist");
+    return 1;
+  }
+
+  string s;
+  map<string, int> counter;
+
+  // While there are lines left to read, read s into a new line
+  while (getline(infile, s)) {
+    vector<string> words = split(lower(s));
+    // Strip punctutation from the next line
+    transform(words.begin(), words.end(), words.begin(), noPunct);
+    // Add all words over 4 letters long to the counter
+    for_each(words.begin(), words.end(), [&counter](string t) {
+      if (t.length() > 4)
+        ++counter[t];
+    });
+  }
+
+  vector<pair<string, int>> sortedCounter = sortedMap(counter);
+  int maxWordLength = maxLen(counter);
+
+  string outname = "output.txt";
+  if (argv == 3) // If the user has given a name for outfile, then use it
+    outname = argc[2];
+
+  ofstream outfile(outname);
+  outfile << "Word" << setw(maxWordLength) << "Usage\n\n";
+
+  // Print each word to outfile, with buffer making sure everything is nicely
+  // aligned
+  for_each(sortedCounter.begin(), sortedCounter.end(),
+           [&](const pair<string, int> &p) {
+	     outfile << p.first << " ";
```

Comment:
I should have made that a bit clearer. Here is an example:
```cpp
  std::cout << "someString" << std::setw(20) << std::setfill(' ') << "andAnotherString" << "\n";
  std::cout << "otherString" << std::setw(20) << std::setfill(' ') << "andAnotherString2" << "\n";
```
this will output:
```
someString    andAnotherString                                                                                                                                                   
otherString   andAnotherString2    
```

You can use the same maxWordLength variable and don't have to subtract each time in the loop. I hope the example clarifies this a bit